### PR TITLE
Replace Water library with Wireguard's tun package (inc. Wintun support on Windows)

### PR DESCRIPTION
**Pull request by neilalexander**
This replaces the `water` library, which has handled TUN/TAP so far, with Wireguard's `tun` library. This PR also effectively removes all TAP support and surrounding code, as TUN will be usable on all supported platforms. 

The most important side effect of this PR is that Wintun is now supported on Windows instead of the hopelessly-awful OpenVPN TAP driver. Performance is *significantly* improved with it and the setup code is much more reliable (no longer relying on `netsh` calls etc).

Note that, in this branch, the OpenVPN TAP driver is **no longer supported** on Windows.

One thing to note is that Wintun is only officially distributed as a signed `.msm` file for integration into an `.msi` installer. Therefore we need to come up with a way to generate MSIs and embed the Wintun driver into it. I had a brief go at this using GNOME's `msitools`/`wixl` but it turns out merging `.msm` files is not supported by that software (missing feature). It'd be great if whatever solution we come up with can also be done for each build using CircleCI so that it is not a manual process to package.

This should address #487 & #471.

### CODE REVIEW

**Comment by neilalexander** [comment]
Test builds of commit bd92c11 are available below. Please install Wireguard for Windows first, as this will ensure that the Wintun driver is installed correctly. The OpenVPN TAP driver is no longer supported in this build.

64-bit: https://2380-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0012-windows-amd64.exe
32-bit: https://2380-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0012-windows-i386.exe

**Comment by neilalexander** [comment]
Test builds of commit bd92c11 for Windows are available below. Please install Wireguard for Windows first, as this will ensure that the Wintun driver is installed correctly. The OpenVPN TAP driver is no longer supported in this build.

64-bit: https://2380-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0012-windows-amd64.exe
32-bit: https://2380-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0012-windows-i386.exe

@Vort, @Revertron — You may be interested in testing this build.

**Comment by Vort** [comment]
I have installed it with `openvpn-install-2.5_git-I607-Win7.exe` - is this ok?
With `yggdrasil-goodbyewater-0.3.11-0043-windows-amd64.exe` with each restart Windows asked for network location.
There are no such dialogs with `yggdrasil-goodbyewater-0.3.12-0012-windows-amd64.exe`. But network names still have numbers: Yggdrasil 6, Yggdrasil 7, ... Do I need to clean something in OS?

**Comment by Vort** [comment]
I have installed Wintun with `openvpn-install-2.5_git-I607-Win7.exe` - is that ok?
I see your commit 0529910b01015332b0568342f82830baf52fe158, but Windows still asks for new network location with each service restart. Right now it is named as "Yggrasil  8". Looks like I will need to clean them up at some time.

**Comment by neilalexander** [comment]
The only Wintun driver version I have tested with is v0.7, so anything that installs that should be fine. Earlier versions might also work fine.

Commit 0529910 reuses the same GUID now so new network locations *shouldn't* be recreated with each startup, but I have only tested this on Windows 10 and Windows Server 2016.

It looks like network profiles are stored in `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\NetworkList\Profiles` - it might be possible to clean them up there.

**Comment by Vort** [comment]
The good thing is that I can make large pings now:
`ping y.thingylabs.io -l 65500`
The bad thing is that I can't recheck packet length with Wireshark: there are no Wintun adapters in the list.
Also now two adapters are installed in my system: Wintun Userspace Tunnel and WireGuard Tunnel. Do I need both of them?
I have also tested out-of-order problem (`iperf3 -6 -c y.thingylabs.io -u`): it is still there, nothing changed.

**Comment by neilalexander** [comment]
Where do you see `Wintun Userspace Tunnel`? I don't think I've come across that yet.

Also I'm glad to hear that the MTU problem is fixed at least—there does still seem to be a reordering problem but as best as I can tell it only happens in one direction. Are you able to verify this by toggling the `-R` flag in `iperf3`?

**Comment by Vort** [comment]
> Where do you see Wintun Userspace Tunnel?

Here:
tun
That may be the side effect of OpenVPN installation.

> It looks like network profiles are stored in HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\NetworkList\Profiles - it might be possible to clean them up there.

I can't see `NetworkList` key in registry.

> there does still seem to be a reordering problem but as best as I can tell it only happens in one direction. Are you able to verify this by toggling the -R flag in iperf3?

No out-of-order with `-R`, that is good.

**Comment by neilalexander** [comment]
> That may be the side effect of OpenVPN installation.

Must be, as I don't see that with a Wireguard installation.

> I can't see `NetworkList` key in registry.

I guess that's another thing that's different with more recent versions of Windows!

> No out-of-order with `-R`, that is good.

That's promising at least. What kind of performance are you seeing and on what types of links?

**Comment by r4sas** [comment]
> Therefore we need to come up with a way to generate MSIs and embed the Wintun driver into it.

That is not required. https://git.zx2c4.com/wintun/about/#installation
We can just provide batch with XML configuration to pass it in `msiexec`.

**Comment by Vort** [comment]
> We can just provide batch with XML configuration to pass it in msiexec.

Did you tried it?

**Comment by Vort** [comment]
Whoops. Looks like we have a problem:
```
2019/11/24 19:49:41 Stopping multicast module
2019/11/24 19:49:41 Error reading TUN: file already closed
2019/11/24 19:49:41 Error reading TUN: file already closed
2019/11/24 19:49:41 Error reading TUN: file already closed
2019/11/24 19:49:41 Error reading TUN: file already closed
...
```
300 megabytes of *that* :(

**Comment by Vort** [comment]
> Must be, as I don't see that with a Wireguard installation.

Yes, `wireguard-amd64-0.0.35.msi` produces no extra device.

**Comment by neilalexander** [comment]
> Whoops. Looks like we have a problem:

Yes, this is a bug I've seen and it's not just on Windows. I have this slight feeling it is also related to the ordering problem, if there's more than one reader running for some reason. We're looking into it now.

**Comment by Vort** [comment]
> What kind of performance are you seeing and on what types of links?

5 Mbits/sec TCP with y.thingylabs.io without -R, 12 Mbits/sec with -R.
30 MBits/sec TCP with my virtual machine, 7 Mbits/sec with -R.
9 MBits/sec TCP from my virtual machine, 26 Mbits/sec with -R.

**Comment by Vort** [comment]
> That's promising at least.

Sadly, that happens because y.thingylabs.io uses Linux version.
VM<-->VM generates out-of-order with and without -R.

**Comment by Vort** [comment]
`yggdrasil-goodbyewater-0.3.12-0013-windows-amd64.exe`:
```
2019/11/24 22:05:12 Stopping multicast module
2019/11/24 22:05:12 Error reading TUN: file already closed
2019/11/24 22:05:12 TUN iface write mismatch: 0 bytes written vs 60 bytes given
2019/11/24 22:05:12 TUN iface write mismatch: 0 bytes written vs 350 bytes given
2019/11/24 22:05:13 TUN iface write mismatch: 0 bytes written vs 1408 bytes given
2019/11/24 22:05:13 TUN iface write error: file already closed
2019/11/24 22:05:13 TUN iface write error: file already closed
2019/11/24 22:05:13 TUN iface write error: file already closed
2019/11/24 22:05:13 Stopping... 
```

**Comment by r4sas** [comment]
I used [example](https://git.zx2c4.com/wintun/tree/msi-example) from *wintun* repository and created new one to build installation MSI: https://github.com/r4sas/yggdrasil-wintun
Build artifacts: https://ci.appveyor.com/project/r4sas/yggdrasil-wintun/build/artifacts

**Comment by r4sas** [comment]
```
>iperf3 -6 -c y.thingylabs.io
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   107 MBytes  89.7 Mbits/sec                  sender
[  4]   0.00-10.00  sec   106 MBytes  88.6 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  60.9 MBytes  51.1 Mbits/sec    0             sender
[  4]   0.00-10.00  sec  60.9 MBytes  51.1 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.26 MBytes  1.06 Mbits/sec  0.812 ms  6/161 (3.7%)
[  4] Sent 161 datagrams

>iperf3 -6 -c y.thingylabs.io -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.06 Mbits/sec  0.683 ms  0/162 (0%)
[  4] Sent 162 datagrams
```

**Comment by neilalexander** [comment]
> I used example from wintun repository and created new one to build installation MSI:

Interesting, I did actually attempt a WiX template: https://github.com/neilalexander/yggdrasil-go/blob/msi/contrib/msi/create-wix.sh

We might be able to easily enough complete the package and then appveyor it into existence.

**Comment by Vort** [comment]
> // FIXME some platforms return the wrong number of bytes written, causing error spam

Looks like this happens only after some errors.
So if you eliminate causes for errors, incorrect result should not appear.

**Comment by neilalexander** [comment]
> Looks like this happens only after some errors.
> So if you eliminate causes for errors, incorrect result should not appear.

This is because of the offset bytes in the `tun` package. I don't think it's a case of there really being an error, it's that the `tun.Write` function on different platforms accounts for the offsets (and possibly Protocol Information) headers differently. The warning we printed was informational only - we never actually acted on it.

**Comment by Vort** [comment]
> This is because of the offset bytes in the `tun` package. I don't think it's a case of there really being an error

Maybe. But 1408 do not looks like offset:
`TUN iface write mismatch: 0 bytes written vs 1408 bytes given`

It looks like software tries to write data to closed device:
`Error reading TUN: file already closed`

And instead of getting `err`, it gets `0` bytes response.

**Comment by neilalexander** [comment]
> It looks like software tries to write data to closed device:

When you stop Yggdrasil, there should always be one read error and there might sometimes be one write error.

This is because a pending read operation that is currently blocked waiting for a packet will fail when the TUN is closed—that is why you see the `file already closed` error. If there's a write taking place in that moment too, that will also error. It's a harmless condition and we can possibly hide the error if it's known that Yggdrasil is already shutting down by that point.

What is incorrect is seeing lots of those errors, but that should be fixed now after 6560aac.

**Comment by neilalexander** [comment]
The packet reordering problem *should* be resolved with 837e7da - it's not a permanent fix but it at least isolates the problem. 

@Vort, @Revertron: Please let me know if the performance is any better with these builds:

64-bit: https://2424-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0020-yggdrasilctl-windows-amd64.exe
32-bit: https://2424-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0020-yggdrasilctl-windows-i386.exe

**Comment by neilalexander** [comment]
Apologies that the previous comment accidentally linked to `yggdrasilctl` instead of `yggdrasil` - I have corrected it now.

**Comment by Vort** [comment]
I have made 4 test runs:
21/241, 3/212, 1/198 and 4/211 out of order packets.
Something is changed, but I can't determine what exactly.

**Comment by Vort** [comment]
I see more differences now. Will update message shortly.

**Comment by Vort** [comment]
When I started to test Host Win7 <=> VM Win7, I have discovered that there are no reordering with UDP test.
But, sadly, connection become unusable. Even simple ping shows the problems:
ping
I saw lags, packet drops and other glitches, which is hard to describe:
udp_test
And when I test VM <=> y.thingylabs.io, I see out of order packets.
So this fix definetly requires more polishing.

**Comment by Vort** [comment]
I think, 837e7da7921d78aa58f6e1a10220184d99d3f6e7 should be reverted. More correct fix is here: 98339cdc3f89cc4342c5e7ef0624dc440ff8a4ac.
With that commit applied, I have such speeds:
`iperf3 -6 -c y.thingylabs.io`: 7 MBits/s
`iperf3 -6 -c y.thingylabs.io -R`: 13 MBits/s
`iperf3 -6 -c <my-vm-ip>`: **375** MBits/s
`iperf3 -6 -c <my-vm-ip> -R`: **423** MBits/s

**Comment by neilalexander** [comment]
Yes, 837e7da will have to be reverted in order for the fix in 98339cd to properly take effect.

**Comment by neilalexander** [comment]
Seemingly with 98339cd but with 837e7da reverted, send performance drops horribly:
```
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-1.01   sec   512 KBytes  4.15 Mbits/sec
[  4]   1.01-2.01   sec  0.00 Bytes  0.00 bits/sec
[  4]   2.01-3.00   sec  0.00 Bytes  0.00 bits/sec
[  4]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec
[  4]   4.00-5.01   sec  0.00 Bytes  0.00 bits/sec
[  4]   5.01-6.01   sec   128 KBytes  1.05 Mbits/sec
[  4]   6.01-7.01   sec  0.00 Bytes  0.00 bits/sec
[  4]   7.01-8.02   sec  0.00 Bytes  0.00 bits/sec
[  4]   8.02-9.00   sec  0.00 Bytes  0.00 bits/sec
[  4]   9.00-10.01  sec  0.00 Bytes  0.00 bits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.01  sec   640 KBytes   524 Kbits/sec                  sender
[  4]   0.00-10.01  sec   428 KBytes   350 Kbits/sec                  receiver
```
Whereas with both 98339cd and 837e7da:
```
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-1.00   sec  46.2 MBytes   387 Mbits/sec
[  4]   1.00-2.00   sec  56.4 MBytes   473 Mbits/sec
[  4]   2.00-3.00   sec  51.1 MBytes   429 Mbits/sec
[  4]   3.00-4.00   sec  48.9 MBytes   410 Mbits/sec
[  4]   4.00-5.00   sec  49.2 MBytes   414 Mbits/sec
[  4]   5.00-6.00   sec  46.5 MBytes   390 Mbits/sec
[  4]   6.00-7.00   sec  48.9 MBytes   410 Mbits/sec
[  4]   7.00-8.00   sec  47.8 MBytes   400 Mbits/sec
[  4]   8.00-9.00   sec  48.1 MBytes   404 Mbits/sec
[  4]   9.00-10.00  sec  50.2 MBytes   421 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   493 MBytes   414 Mbits/sec                  sender
[  4]   0.00-10.00  sec   493 MBytes   414 Mbits/sec                  receiver
```

**Comment by Revertron** [comment]
> Whereas with both [98339cd](https://github.com/yggdrasil-network/yggdrasil-go/commit/98339cdc3f89cc4342c5e7ef0624dc440ff8a4ac) and [837e7da](https://github.com/yggdrasil-network/yggdrasil-go/commit/837e7da7921d78aa58f6e1a10220184d99d3f6e7) in the tree:
> 
> [ ID] Interval           Transfer     Bandwidth
> [  4]   0.00-1.00   sec  46.2 MBytes   387 Mbits/sec
> [  4]   1.00-2.00   sec  56.4 MBytes   473 Mbits/sec

And what MTU do you use in these tests?

**Comment by neilalexander** [comment]
Looks like something in the most recent `goodbyewater` builds has completely killed things, hmm.

> And what MTU do you use in these tests?

Session MTU is 9000.

**Comment by neilalexander** [comment]
OK, I've reverted 837e7da since, with a direct peering, the performance is fine without it in both directions. The slowness I was experiencing must have been down to the intermediary node, which I will explore more separately.

**Comment by Vort** [comment]
@Revertron, @r4sas:
Can you test [yggdrasil-goodbyewater-0.3.12-0028-windows-amd64.exe](https://2440-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0028-windows-amd64.exe)?
It looks good for me. Only problem I see is with `-u -R` test from host to VM. And it is not reordering.

**Comment by Revertron** [comment]
Yes, the build 28 is a big leap forward. It gives me about 800MB/s download and 600MB/s upload in LAN. But the UDP is slow indeed - about 1.5MB/s max.
P.S.: The best MTU for me is 35400 for now :)

**Comment by Vort** [comment]
UDP speed can be controlled by parameters of `iperf3`. But that is another topic :)

**Comment by r4sas** [comment]
> Can you test [yggdrasil-goodbyewater-0.3.12-0028-windows-amd64.exe](https://2440-115685026-gh.circle-artifacts.com/0/yggdrasil-goodbyewater-0.3.12-0028-windows-amd64.exe)?

```
>iperf3 -6 -c y.thingylabs.io
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  25.5 MBytes  21.4 Mbits/sec                  sender
[  4]   0.00-10.00  sec  23.5 MBytes  19.7 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  7.18 MBytes  6.03 Mbits/sec   17             sender
[  4]   0.00-10.00  sec  5.25 MBytes  4.40 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.06 Mbits/sec  1.498 ms  0/162 (0%)
[  4] Sent 162 datagrams

>iperf3 -6 -c y.thingylabs.io -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.06 Mbits/sec  28.543 ms  23/162 (14%)
[  4] Sent 162 datagrams
[SUM]  0.0-10.0 sec  23 datagrams received out-of-order
```

**Comment by Vort** [comment]
Do you have direct peering with `ygg.thingylabs.io`?

**Comment by r4sas** [comment]
> Do you have direct peering with `ygg.thingylabs.io`?

No. Configuration is default, with only one TLS ipv4/v6 peer and without filled Listen option.
But I think daemon automaticly connects to that peer.

**Comment by Vort** [comment]
> Configuration is default, with only one TLS ipv4/v6 peer and without filled Listen option.

This means that your traffic can go different ways with each test run.
Such tests are valuable too, but are less stable.

**Comment by neilalexander** [comment]
To test raw peering performance, please try configuring exactly one peer entry, rather than IPv4+IPv6 connectivity to the same peer, as that will cause switch behaviours to kick in.

**Comment by r4sas** [comment]
ok, direct with one only peer entry to thingylabs:

```
>iperf3 -6 -c y.thingylabs.io
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   106 MBytes  88.8 Mbits/sec                  sender
[  4]   0.00-10.00  sec   104 MBytes  87.5 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  15.9 MBytes  13.4 Mbits/sec    0             sender
[  4]   0.00-10.00  sec  15.0 MBytes  12.6 Mbits/sec                  receiver

>iperf3 -6 -c y.thingylabs.io -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.07 Mbits/sec  17.333 ms  0/163 (0%)
[  4] Sent 163 datagrams

>iperf3 -6 -c y.thingylabs.io -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.06 Mbits/sec  31.419 ms  0/162 (0%)
[  4] Sent 162 datagrams
```

**Comment by Revertron** [comment]
@r4sas can you try with `201:2522::5736:9a6:7dd1:7c95` please?

**Comment by r4sas** [comment]
> @r4sas can you try with `201:2522::5736:9a6:7dd1:7c95` please?

Connection refused.

**Comment by Revertron** [comment]
@r4sas Sorry, I've got Internet accident, you can try now.

**Comment by r4sas** [comment]
```
### My TLS peer:
>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  33.1 MBytes  27.8 Mbits/sec                  sender
[  4]   0.00-10.00  sec  29.5 MBytes  24.8 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  8.18 MBytes  6.86 Mbits/sec    5             sender
[  4]   0.00-10.00  sec  6.87 MBytes  5.76 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.26 MBytes  1.06 Mbits/sec  16.177 ms  0/161 (0%)
[  4] Sent 161 datagrams

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.07 Mbits/sec  13.345 ms  0/163 (0%)
[  4] Sent 163 datagrams

### Thingylabs TCP:
>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec   100 MBytes  84.1 Mbits/sec                  sender
[  4]   0.00-10.00  sec  98.9 MBytes  82.9 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  11.4 MBytes  9.53 Mbits/sec    0             sender
[  4]   0.00-10.00  sec  9.37 MBytes  7.86 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.29 MBytes  1.08 Mbits/sec  14.523 ms  0/165 (0%)
[  4] Sent 165 datagrams

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.07 Mbits/sec  5.673 ms  0/163 (0%)
[  4] Sent 163 datagrams
```

**Comment by Revertron** [comment]
@r4sas what MTU are you using? The numbers are strange :(

**Comment by r4sas** [comment]
> @r4sas what MTU are you using? The numbers are strange :(

All settings are default, it must be 65535.

**Comment by Revertron** [comment]
> > @r4sas what MTU are you using? The numbers are strange :(
> 
> All settings are default, it must be 65535.

Try 35400 please ;)

**Comment by r4sas** [comment]
> > > @r4sas what MTU are you using? The numbers are strange :(
> > 
> > 
> > All settings are default, it must be 65535.
> 
> Try 35400 please ;)

```
### My TCP peer:
>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  81.4 MBytes  68.3 Mbits/sec                  sender
[  4]   0.00-10.00  sec  79.9 MBytes  67.1 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  21.7 MBytes  18.2 Mbits/sec    3             sender
[  4]   0.00-10.00  sec  21.2 MBytes  17.8 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.07 Mbits/sec  1.688 ms  0/163 (0%)
[  4] Sent 163 datagrams

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.28 MBytes  1.07 Mbits/sec  6.793 ms  0/164 (0%)
[  4] Sent 164 datagrams

### Thingylabs TCP:
>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  99.6 MBytes  83.6 Mbits/sec                  sender
[  4]   0.00-10.00  sec  98.2 MBytes  82.4 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -R
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  6.56 MBytes  5.50 Mbits/sec    0             sender
[  4]   0.00-10.00  sec  5.47 MBytes  4.59 Mbits/sec                  receiver

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.27 MBytes  1.07 Mbits/sec  10.379 ms  0/163 (0%)
[  4] Sent 163 datagrams

>iperf3 -6 -c 201:2522::5736:9a6:7dd1:7c95 -u -R
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.00  sec  1.28 MBytes  1.07 Mbits/sec  19.454 ms  0/164 (0%)
[  4] Sent 164 datagrams
```

**Comment by Revertron** [comment]
Okay, I think this is the best you can get from your location.

**Comment by Vort** [comment]
By the way, thingylabs have iperf3 service on both interfaces.
So you can compare Yggdrasil link with direct connection.
Here is what I have now:
`iperf3 -6 -c y.thingylabs.io`, `IfMTU: 65535`: 8 Mbits/sec
`iperf3 -6 -c y.thingylabs.io`, `IfMTU: 1400`: 20 Mbits/sec
`iperf3 -4 -c ygg.thingylabs.io`: 38 Mbits/sec
Looks like there is a space for improvements at Yggdrasil side.

**Comment by Vort** [comment]
I have found where it is possible to remove extra network locations in Windows 7:
win7_network_locations

**Comment by neilalexander** [comment]
@Vort, @Revertron, @r4sas I have created some MSIs using Appveyor. For some reason the build shows failed in this PR and I'll fix that tomorrow, but the following should be the equivalent build:

64-bit: https://ci.appveyor.com/api/buildjobs/xl5jy0mu6awb70vy/artifacts/yggdrasil-goodbyewater-0.3.12-0029-x64.msi
32-bit: https://ci.appveyor.com/api/buildjobs/xl5jy0mu6awb70vy/artifacts/yggdrasil-goodbyewater-0.3.12-0029-x86.msi

Please let me know if they work.

**Comment by Vort** [comment]
So many things needs to be done here...
1. Service is added with `-autoconf`. So generating of config does nothing.
2. Writing to subfolders of `Program Files` requires admin rights, which may be a problem in some cases.
3. It is not possible right now to customize installation path, which is bad for me, since I store my software at disk `D:` (but it is ok for driver however).

I hope that `.msi` will be not the only installation optoin available, at least for now.

**Comment by neilalexander** [comment]
Yes I'm about to fix the `-autoconf` bit.

Admin privileges are required to install Wintun anyway, and they are also required for adding the Windows service. I am happy that admin privileges are required to install.

As for the install path, that's probably simple enough to do - I just haven't got to it yet.

**Comment by neilalexander** [comment]
OK, this build should use `-useconffile` properly now:

64-bit: https://ci.appveyor.com/api/buildjobs/la9wwvlsd2c9676r/artifacts/yggdrasil-goodbyewater-0.3.12-0032-x64.msi
32-bit: https://ci.appveyor.com/api/buildjobs/la9wwvlsd2c9676r/artifacts/yggdrasil-goodbyewater-0.3.12-0032-x86.msi

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  do_not_increment_build_number: true
+os: Visual Studio 2017
+shallow_clone: false
+
+environment:
+  MSYS2_PATH_TYPE: inherit
+  CHERE_INVOKING: enabled_from_arguments
+
+install:
+- c:\msys64\usr\bin\bash -lc "pacman --noconfirm -S curl unzip"
```
Comment (r4sas) [inline appveyor.yml]: You can skip that part if preinstall called in script (line 32).

Comment (neilalexander) [reply appveyor.yml]: Cleaned it up.

**Pull request merged**